src/mlpack/methods/ann/layer/group_norm_impl.hpp

```diff
@@ -0,0 +1,186 @@
+  beta = arma::mat(weights.memptr() + gamma.n_elem, size, 1, false, false);
+
+  if (!loading)
+  {
+    gamma.fill(1.0);
+    beta.fill(0.0);
+  }
+
+  loading = false;
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void GroupNorm<InputDataType, OutputDataType>::Forward(
+    const arma::Mat<eT>& input, arma::Mat<eT>& output)
+{
+  assert(input.n_rows % channelSize == 0);
+  assert(input.n_rows % size == 0);
+  arma::mat reshapedInput(const_cast<arma::Mat<eT>&>(input).memptr(),
+    input.n_rows * groupCount / size, input.n_cols * size / groupCount, false, false);
+
+  if (output.is_empty())
+    output.zeros(input.n_rows, input.n_cols);
+
+  arma::mat reshapedOutput((output).memptr(),
+    input.n_rows * groupCount / size, input.n_cols * size / groupCount, false, false);
+
+  mean = arma::mean(reshapedInput, 0);
+  variance = arma::var(reshapedInput, 1, 0);
+
+  // Normalize the input.
+  reshapedOutput += reshapedInput.each_row() - mean;
+  inputMean = reshapedOutput;
+  reshapedOutput.each_row() /= arma::sqrt(variance + eps);
+
+  // Reused in the backward and gradient step.
+  normalized = output;
+
+  arma::mat expandedGamma, expandedBeta;
+  expandedGamma.set_size(output.n_rows, 1);
+  expandedBeta.set_size(output.n_rows, 1);
+  for (size_t r = 0; r < output.n_rows; ++r)
+  {
+    expandedGamma(r) = gamma(r % size);
+    expandedBeta(r) = beta(r % size);
+  }
+
+  // Scale and shift the output.
+  output.each_col() %= expandedGamma;
+  output.each_col() += expandedBeta;
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void GroupNorm<InputDataType, OutputDataType>::Backward(
+    const arma::Mat<eT>& input, const arma::Mat<eT>& gy, arma::Mat<eT>& g)
+{
+  arma::mat inputReshaped(const_cast<arma::Mat<eT>&>(input).memptr(),
+    size / groupCount, groupCount * input.n_cols, false, false);
+  arma::mat gReshaped((g).memptr(),
+    size / groupCount, groupCount * g.n_cols, false, false);
+
+  arma::mat expandedGamma;
+  expandedGamma.set_size(input.n_rows, 1);
+  for (size_t r = 0; r < input.n_rows; ++r)
+  {
+    expandedGamma(r) = gamma(r % size);
+  }
+
+  // dl / dxhat.
+  const arma::mat norm = gy.each_col() % expandedGamma;
+
+  arma::mat normReshaped(const_cast<arma::Mat<eT>&>(norm).memptr(),
+    size / groupCount, groupCount * gy.n_cols, false, false);
+
+
+  const arma::mat stdInv = 1.0 / arma::sqrt(variance + eps);
+
+  // sum dl / dxhat * (x - mu) * -0.5 * stdInv^3.
+  const arma::mat var = arma::sum(normReshaped % inputMean, 0) %
+      arma::pow(stdInv, 3.0) * -0.5;
+
+  // dl / dxhat * 1 / stdInv + variance * 2 * (x - mu) / m +
+  // dl / dmu * 1 / m.
+  gReshaped = (normReshaped.each_row() % stdInv) + (inputMean.each_row() %
+      var * 2 / inputReshaped.n_rows);
+
+  // sum (dl / dxhat * -1 / stdInv) + variance *
+  // (sum -2 * (x - mu)) / m.
+  gReshaped.each_row() += arma::sum(normReshaped.each_row() % -stdInv, 0) / inputReshaped.n_rows;
```

Comment:
```suggestion
  gReshaped.each_row() += arma::sum(normReshaped.each_row() % -stdInv, 0) /
      inputReshaped.n_rows;
```

To keep the line length.